distributor.go

```diff
@@ -176,53 +184,87 @@ func (d *Distributor) Append(ctx context.Context, samples []*model.Sample) error
 
 	d.receivedSamples.Add(float64(len(samples)))
 
-	samplesByIngester := map[string][]*model.Sample{}
-	for _, sample := range samples {
-		key := tokenForMetric(userID, sample.Metric)
-		ingesters, err := d.cfg.Ring.Get(key, d.cfg.ReplicationFactor)
-		if err != nil {
-			return err
+	sampleTrackers := make([]sampleTracker, len(samples), len(samples))
+	samplesByIngester := map[string][]*sampleTracker{}
+	keys := make([]uint32, len(samples), len(samples))
+
+	for i, sample := range samples {
+		keys[i] = tokenForMetric(userID, sample.Metric)
+	}
+
+	ingesters, err := d.cfg.Ring.BatchGet(keys, d.cfg.ReplicationFactor, ring.Write)
+	if err != nil {
+		return err
+	}
+
+	for i := range samples {
+		sampleTrackers[i] = sampleTracker{
+			sample:     samples[i],
+			minSuccess: (len(ingesters[i]) / 2) + 1,
```

Comment:
Probably worth adding a comment saying the reasoning behind this exact calculation. I guess this also deprecates the `MinWriteSuccess` configuration parameter, so please delete it or add a deprecation warning or similar.